File: sol/solution/0001-0100/0023/Solution.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).
-spec merge_k_lists(Lists :: [#list_node{} | null]) -> #list_node{} | null.
merge_k_lists([]) -> null;
merge_k_lists(Lists) -> merge_all(Lists).
merge_all([L]) -> L;
merge_all(Lists) -> merge_all_pairs(Lists, []).
merge_all_pairs([], Acc) ->
    merge_all(lists:reverse(Acc));
merge_all_pairs([A], Acc) ->
    merge_all(lists:reverse([A | Acc]));
merge_all_pairs([A, B | Rest], Acc) ->
    merge_all_pairs(Rest, [merge_two(A, B) | Acc]).
merge_two(null, B) -> B;
merge_two(A, null) -> A;
merge_two(#list_node{val = Va, next = Na} = A,
          #list_node{val = Vb, next = Nb} = B) ->
    if
        Va =< Vb ->
            #list_node{val = Va, next = merge_two(Na, B)};
        true ->
            #list_node{val = Vb, next = merge_two(A, Nb)}
    end.

File: sol/solution/0001-0100/0023/Solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists(lists)
    return nil if lists.nil? || lists.empty?
    min_heap = []
    lists.each_with_index do |list, index|
        min_heap.push([list.val, index]) if list
    end
    min_heap.sort_by!(&:first)
    dummy = ListNode.new
    current = dummy
    while !min_heap.empty?
        val, index = min_heap.shift
        current.next = ListNode.new(val)
        current = current.next
        lists[index] = lists[index].next if lists[index]
        if lists[index]
            min_heap.push([lists[index].val, index])
            min_heap.sort_by!(&:first)
        end
    end
    dummy.next
end